laurentbartholdi-argp

laurentbartholdi opened the pull request:
This compiles the library Argp-standalone for argument parsing

imciner2 commented on A/Argp/build_tarballs.jl:
```diff
@@ -0,0 +1,44 @@
+
+autoreconf -i
+./configure --prefix=$prefix --build=${MACHTYPE} --host=${target}
+make CC=gcc
```
```suggestion
make CC=${CC}
```
Don't hardcode GCC. We prefer to use Clang on some platforms where that is the better supported compiler (e.g. macOS).

laurentbartholdi replied:
Well, I had initially let the system decide, and compilation breaks with clang. That's why I forced it to gcc.

imciner2 replied:
What were the errors?

imciner2 commented on A/Argp/build_tarballs.jl:
```diff
@@ -0,0 +1,44 @@
+autoreconf -i
+./configure --prefix=$prefix --build=${MACHTYPE} --host=${target}
+make CC=gcc
+install argp.h $includedir
+install libargp.a $libdir
```
```suggestion
install -Dvm 644 argp.h $includedir
install -Dvm 644 libargp.a $libdir
```

imciner2 commented on A/Argp/build_tarballs.jl:
```diff
@@ -0,0 +1,44 @@
+platforms = supported_platforms()
+
+# The products that we will ensure are always built
+products = Product[
+]
```
```suggestion
products = Product[
    # argp is meant as a build-time dependency to be included in other recipes, so static library only
    FileProduct("lib/libargp.a", :libargp),
    FileProduct("include/argp.h", :argp_h),
]
```
It is useful to add these as file products so that way the auditor pass will fail if they accidentally change names or go missing.

imciner2 commented on A/Argp/build_tarballs.jl:
```diff
@@ -0,0 +1,44 @@
+cd ${WORKSPACE}/srcdir/argp-standalone
+
+for f in ${WORKSPACE}/srcdir/patches/*.patch; do
+    atomic_patch -p1 ${f}
```
Are these patches upstreamed or going to be upstreamed? It would be great if they can be incorporated into upstream so we don't have to continually carry them and update them on new versions.

laurentbartholdi replied:
I think this library has not undergoing any development for the last 10 years; one of the patches is for compilation on mingw. I don't plan to upstream them, if that's not too much of a disappointment to you :)

imciner2 commented on A/Argp/build_tarballs.jl:
```diff
@@ -0,0 +1,44 @@
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.9")
```
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6")
```
Do you need a Julia 1.9 compat? This looks simple enough that we could get by with a 1.6 compat.

laurentbartholdi commented:
BTW @imciner2 thanks for suggesting the "products"... it turns out my install command was wrong in 2 ways: it created a file $destdir/include rather than a directory, and it put the libargp.a in $librarydir = $destdir/bin rather than $destdir/lib

laurentbartholdi commented:
The darwin build, using cc, misses"mempcpy" and "strchrnul". Since they're
not standard functions, and the build works with gcc, I'd rather not open a
can of worms.

On Mon, 5 Aug 2024 at 15:25, Ian McInerney ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In A/Argp/build_tarballs.jl
> <https://github.com/JuliaPackaging/Yggdrasil/pull/9197#discussion_r1704118437>
> :
>
> > +sources = [
> +    GitSource("https://github.com/ericonr/argp-standalone.git", "743004c68e7358fb9cd4737450f2d9a34076aadf"),
> +    DirectorySource("./bundled")
> +]
> +
> +# Bash recipe for building across all platforms
> +script = raw"""
> +cd ${WORKSPACE}/srcdir/argp-standalone
> +
> +for f in ${WORKSPACE}/srcdir/patches/*.patch; do
> +    atomic_patch -p1 ${f}
> +done
> +
> +autoreconf -i
> +./configure --prefix=$prefix --build=${MACHTYPE} --host=${target}
> +make CC=gcc
>
> What were the errors?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaPackaging/Yggdrasil/pull/9197#discussion_r1704118437>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARAQUDKADGYTNVSG6KO4YDZP54K5AVCNFSM6AAAAABL7W2QJCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDEMJZGAYDAOJRGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Laurent Bartholdi     laurent.bartholdi<at>gmail<dot>com
Fachrichtung Informatik+Mathematik, Universität des Saarlandes
Postfach 151150, 66041 Saarbrücken, Germany
Tel. +49 681 3023227, Sekr. +49 681 3023430

imciner2 commented:
The problem is, this is creating a static library, and I don't think there are any guarantees about how compatible static libraries are across compiler versions/different compilers. A static library is just an archive file containing the individual object files that haven't been linked together, so they may contain things that other compilers don't support. That is why we prefer to keep everything using one compiler on a platform - so we can avoid problems due to this.

laurentbartholdi commented:
OK, it wasn't so bad. I added missing prototypes and now it seems to work.

Since you're an expert, would you mind giving a look to my next recipe
https://github.com/JuliaPackaging/Yggdrasil/pull/9198 ? When I compile on
my macos laptop it works fine, but when I cross-compile for
aarch64-apple-darwin20 I get, at the link phase,

[15:18:22] /bin/sh ../libtool --tag=CXX --mode=link clang++ -O2 -Wall
-DNDEBUG -UDEBUG -version-info 11:1:2 -no-undefined -o libgivaro.la -rpath
/workspace/destdir/lib dummy.lo kernel/gmp++/libgmppp.la kernel/bstruct/
libgivbstruct.la kernel/integer/libgivinteger.la kernel/memory/
libgivmemory.la kernel/rational/libgivrational.la kernel/system/
libgivsystem.la kernel/field/libgivfield.la library/tools/libgivtools.la
library/poly1/libgivpoly1.la -lgmpxx -lgmp [15:18:23] libtool: link: c++ -r
-keep_private_externs -nostdlib -o .libs/libgivaro.9.dylib-master.o
.libs/dummy.o -Wl,-force_load,kernel/gmp++/.libs/libgmppp.a
-Wl,-force_load,kernel/bstruct/.libs/libgivbstruct.a
-Wl,-force_load,kernel/integer/.libs/libgivinteger.a
-Wl,-force_load,kernel/memory/.libs/libgivmemory.a
-Wl,-force_load,kernel/rational/.libs/libgivrational.a
-Wl,-force_load,kernel/system/.libs/libgivsystem.a
-Wl,-force_load,kernel/field/.libs/libgivfield.a
-Wl,-force_load,library/tools/.libs/libgivtools.a
-Wl,-force_load,library/poly1/.libs/libgivpoly1.a [15:18:23] ld64.lld:
warning: Option `-keep_private_externs' is not yet implemented. Stay
tuned... [15:18:23] ld64.lld: warning: Option `-r' is not yet implemented.
Stay tuned... [15:18:23] ld64.lld: warning: Option `-r' is not yet
implemented. Stay tuned... [15:18:23] ld64.lld: error: undefined symbol:
main [15:18:23] >>> referenced by the entry point [15:18:23] [15:18:23]
ld64.lld: error: undefined symbol: __cxa_atexit [15:18:23] >>> referenced
by kernel/gmp++/.libs/libgmppp.a(gmp++_int_lib.o):(symbol
_GLOBAL__sub_I_gmp___int_lib.C+0x94) [15:18:23] >>> referenced by
kernel/gmp++/.libs/libgmppp.a(gmp++_int_lib.o):(symbol
_GLOBAL__sub_I_gmp___int_lib.C+0x64) [15:18:23] >>> referenced by
kernel/gmp++/.libs/libgmppp.a(gmp++_int_lib.o):(symbol
_GLOBAL__sub_I_gmp___int_lib.C+0x40) [15:18:23] >>> referenced 6 more times
[15:18:23] [15:18:23] ld64.lld: error: undefined symbol: __gmpz_tdiv_q
[15:18:23] >>> referenced by
kernel/gmp++/.libs/libgmppp.a(gmp++_int_div.o):(symbol
Givaro::Integer::divin(Givaro::Integer&, Givaro::Integer const&)+0x28)
[15:18:23] >>> referenced by
kernel/gmp++/.libs/libgmppp.a(gmp++_int_div.o):(symbol
Givaro::Integer::div(Givaro::Integer&, Givaro::Integer const&,
Givaro::Integer const&)+0x50) [15:18:23] >>> referenced by
kernel/gmp++/.libs/libgmppp.a(gmp++_int_div.o):(symbol
Givaro::Integer::operator/=(Givaro::Integer const&)+0x28) [15:18:23] >>>
referenced 7 more times [15:18:23] [15:18:23] ld64.lld: error: undefined
symbol: _Unwind_Resume ...

which seems to be two errors:
1) libtool assumes that c++ knows some options which it doesn't
2) libtool splits the process into creating
.libs/libgivaro.9.dylib-master.o and then a .dylib, but in the first step
it gets confused and tries to create an executable, which does'nt have any
library (libgmp, libc, libc++).

Again, things work better in Yggdrasil / codespaces if I force CXX=g++: the
link is attempted correctly, and the "only" error is that libgmpxx was
compiled for clang++ which is not compatible with g++ so I get 'undefined
reference to __XXXX_mpz_struct'.

TIA!

On Mon, 5 Aug 2024 at 17:19, Ian McInerney ***@***.***> wrote:

> The problem is, this is creating a static library, and I don't think there
> are any guarantees about how compatible static libraries are across
> compiler versions/different compilers. A static library is just an archive
> file containing the individual object files that haven't been linked
> together, so they may contain things that other compilers don't support.
> That is why we prefer to keep everything using one compiler on a platform -
> so we can avoid problems due to this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaPackaging/Yggdrasil/pull/9197#issuecomment-2269328341>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARAQUGAMN4L2F673ORL7YDZP6JYZAVCNFSM6AAAAABL7W2QJCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENRZGMZDQMZUGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Laurent Bartholdi     laurent.bartholdi<at>gmail<dot>com
Fachrichtung Informatik+Mathematik, Universität des Saarlandes
Postfach 151150, 66041 Saarbrücken, Germany
Tel. +49 681 3023227, Sekr. +49 681 3023430

imciner2 commented:
> Uh, I just noticed that there's already a recipe "argp_standalone". I had missed it because it was all in lowercaps. Mine is better :) because they don't build for apple or mingw.

Please open a PR updating that build script to work on the platforms you need and we'll keep that package instead of adding another one.

laurentbartholdi commented:
Replaced by https://github.com/JuliaPackaging/Yggdrasil/pull/9216
